Optimize and Cleanup CScript::FindAndDelete

pstratem opened the pull request:
This PR improves the worst-case behavior of CScript::FindAndDelete.

I have emphasized the obvious correctness of the algorithm in this commit.

Additionally I have done extensive fuzzing to ensure the result is identical to the current implementation.

gavinandresen committed:
Unit test for CScript::FindAndDelete

kazcw commented:
Is this really an optimization? It increases copying in the non-multisig case, which is vastly more common...

Commit:
Replace memcmp with std::equal in CScript::FindAndDelete

Function is stl; std::equal just makes more sense.

Commit:
Replace c-style cast with c++ style static_cast.

Commit:
Improve worst-case behavior of CScript::FindAndDelete

Thanks to Sergio Lerner for identifying this issue and suggesting this kind of solution.

TheBlueMatt commented:
I highly prefer this to #7884...the original is kinda hard to read, but #7884 is impossible to reasonably convince yourself works, even if it probably does.
I benchmarked syncing on tmpfs with the final call to VerifySignature in CheckSig() disabled on this and #7884 and couldn't detect any difference in sync time through about 280k (just happened to be some blk* flies I had lying around went until there).

TheBlueMatt commented on src/script/script.h:
```diff
@@ -570,17 +570,25 @@ class CScript : public CScriptBase
         while (GetOp(pc, opcode));
+
+        result.insert(result.end(), pc2, end());
+        *this = result;
```
You should gate this and the last insert on an if(nFound)

dcousens commented:
utACK 6443169,  but I'd prefer benchmarks before merge.

sipa commented:
@kazwc @dcousens It is only intended to improve the worst case behaviour, and @TheBlueMatt already benchmarked that it does not measurably affect the average case.

@pstratem If, after includong Matt's suggestion, you drop the result.reserve() call, I think this will run with zero heap effect for non-match situations.

pstratem commented:
@sipa indeed you're right I've made that change

dcousens commented:
@sipa if we can't measure this is an improvement,  isn't this just all speculation?

dcousens commented:
utACK da1d7775,  looks good to merge

laanwj commented:
utACK d1d7775

sipa commented:
@dcousens it gives a very measurable improvement in worst case situations.
Just not in the average case you'll observe with a random benchmark.

sipa commented:
utACK d1d7775587473410a107e7079616b9ecaae8dd06

kellwin13 commented on src/test/script_tests.cpp:
```diff
@@ -1051,4 +1051,121 @@ BOOST_AUTO_TEST_CASE(script_GetScriptAsm)
+    BOOST_CHECK(s == expect);
+}
+
 BOOST_AUTO_TEST_SUITE_END()
```
kellwin
